### ISSUE TITLE
More general proofs by induction in SAW

### ISSUE THREAD

**Issue by eddywestbrook** [GaloisInc/saw-script#1800]
We have recently been trying to write more proofs about folds over cryptol sequences, and some of these require induction over the sequence. As a very simple example, imagine we want to prove

```
map (\x -> x) xs == xs
```

The most straightforward way to prove this fact is by induction on either the structure of `xs` or on its length. Unfortunately, the only induction principle currently available in SAW is `prove_by_bv_induction`, which builds a proof by strong induction on the less-then operation on bitvectors, and so it not relevant here.

At the very least, it would be good to provide a `prove_by_nat_induction`, that would allow proofs by structural induction on a natural number. (We actually *don't* want strong induction, where we assume a property holds for all smaller natural numbers, since the structural induction is much more natural to use here.) Ideally, though, we would provide a more general approach to proof by induction, where the user could provide a more general induction scheme.

One possibility would be a command `prove_by_struct_induction`, that would prove a universal property over an inductive type by applying the recursor for that datatype. This would allow properties like the above to be proved by induction on the length of `xs`, though not by induction on `xs` itself because the vector type is not an inductive type in SAW core. More generally, though, it might be nice to allow an expert user to add more induction principles, such as induction on vectors, that could be used in these sorts of proofs.

**Comment by eddywestbrook** [comment]
Upon further reflection, the difficulty with a more general induction technique like what was outlined above would actually be writing the induction hypothesis in cryptol. You can't quantify over natural numbers in cryptol; instead, you quantify over `Num`, which includes infinity. This quantification also has to be at the type level in cryptol, not the term level.

A solution would be to define a `prove_by_num_induction` tactic, that takes a polymorphic cryptol term of type `{n} t1 -> ... -> tn -> Bit` to describe a predicate over `Num` type index `n`. The tactic would then take 3 proof scripts, one to prove  the case where `n=0`, one for `n=m+1` for some `m` where we have assumed the property holds, and one for `n=infinity`.

In practice, we might often want to use this tactic to prove properties that only hold for finite `Num`s, i.e., for natural numbers. This could be done by using a predicate of the form `{n} fin n => t1 -> ... -> tn -> t`, and using an EFQ-type rule to eliminate the infinity case by introducing and then eliminating the `fin infinity` assumption to the false proposition. Presumably we would want a special form of the `prove_by_num_induction` tactic that does this for the user, though this would hopefully be something that could be written in SAW script as a user-defined function rather than as a distinct tactic in the interpreter.

**Comment by robdockins** [comment]
Certainly, more ways to perform induction proofs would be useful.  However, note that the `Vec` type in SAWCore is a primitive, and not defined inductively, so I don't think an induction proof would help here. Instead, some extensionality axiom is probably needed, or reasoning principles specific to `foldr`, etc.

**Comment by eddywestbrook** [comment]
You're right, we would need some extensionality axioms to prove that any `Vec (S n) a` equals the cons of its head and tail and that any `Vec 0 a` equals `EmptyVec`. But I think we could use these extensionality axioms to add casts in the places where an inductive definition of `Vec` would be able to use definitional equality. That's what would be needed.
